lib/SILOptimizer/Mandatory/TFDevicePartition.cpp

```diff
@@ -37,16 +37,20 @@ using namespace tf;
 // Device Partitioning Utilities
 //===----------------------------------------------------------------------===//
 
-// Return the device attribute associated with `inst`, which is required to
-// exist.
+// Return the device attribute associated with `inst`, if any. Otherwise return
+// an empty string.
 StringRef swift::tf::getDeviceString(const GraphOperationInfo &graphOpInfo) {
   auto attr = graphOpInfo.getInst()->getAttributeNamed(TF_DEVICE_ATTR);
-  assert(attr.hasValue() && "Tensor op instruction has no device string");
-  return attr.getValue().getStringValue();
+  if (attr.hasValue())
+    return attr.getValue().getStringValue();
+  return StringRef();
 }
 
 DeviceId swift::tf::getDeviceId(const GraphOperationInfo &graphOpInfo) {
-  return getOpDeviceId(getDeviceString(graphOpInfo));
+  auto deviceStr = getDeviceString(graphOpInfo);
+  if (!deviceStr.empty())
+    return getOpDeviceId(deviceStr);
```

Comment:
It looks like runtime device type is handled by implicitly having an empty string (here and in other places). Why not make it explicit and expand getOpDeviceId (and other functions) to deal with "RUNTIME" device strings as well?

Reply:
I feel using empty string is somewhat simpler -- when a device string is set, we expect it to be a valid TF device string, which can be passed into graph node creation; otherwise, we expect the op to be "unplaced" (aka "runtime-placed"). In both cases, graph lowering then just get the output of `getDeviceString()`, and use that to set an op device attr(see the call-sites of `getDeviceString()` in TFLowerGraph.cpp).

If instead in the latter case, we use a different string like "RUNTIME", we'll need to do additional interpretation on that when generating a graph node, and that'd add code complexity.

Note this is different from the "RUNTIME" name mangling in the generated graph function name -- there we use the string "RUNTIME" to invoke the function accordingly.

WDYT?

Reply:
I don't have a strong preference, but if it is a trivial change, having a non-empty string would be good. e.g., we will be able to see it in debug messages. 

Don't we have to do something about empty string in TFLowerGraph anyway? I see that `thisDeviceIdStr` is set using the result of getDeviceString() in the constructor of [TFGraphFunctionLowering](https://github.com/apple/swift/blob/359de5b28af78236a97311db8aa3a8d4f1442400/lib/SILOptimizer/Mandatory/TFLowerGraph.cpp#L284) and used in other places. Can't we set up thisDeviceIdStr to the correct value appropriately in the constructor?

Reply:
Debugging should not be a concern. The `DeviceId` struct has a lot of debugging support (e.g. in gdb, or print the content via `dump()` / `print()`).

It's possible to return some string and then translate it back to empty string in graph lowering, but that's extra complexity that I'd like to avoid.

If the choice of the empty string turns out to affect code reading / debugging, I'd be happy to revisit this design decision.